[WIP] Exploration update? Exploration update! (1.11)

aramperes opened the pull request:
Let's start updating Glowstone to 1.11:

- [x] Protocol changes
- [x] Metadata update
  - [x] Add new Metadata entries
  - [x] Implement new entries once API is updated and implemented
- [x] Bukkit, SpigotAPI and Paper update to 1.11
- [x] Deploy Glowkit 1.11
- [x] Implement new Entity classes
  - [x] Horse classes
  - [x] Skeleton classes
  - [x] Zombie classes
  - [x] Tweak entity storage
  - [x] New entities: Vex, Evoker, Vindicator
- [x] Misc changes for the update

-----------

Other stuff:
 - Update the following libraries:
   - `sqlite-jdbc` 3.15.1
   - `mysql-connector-java` 5.1.39
   - Remove `commons-io`
 - Ambient sounds for all living entities

aramperes committed:
Procotol/Metadata update for 1.11

mastercoms commented:
This pull request will not be accepted until you come up with an original title. #293

phase commented:
I can't wait to see `1.12? 1.12!`

aramperes committed:
Update circle.yml to use Glowkit 1.11 branch

aramperes committed:
Don't clone recursively

mastercoms committed:
Use Glowkit 1.11

aramperes committed:
Add new entity classes to metadata entries

aramperes committed:
Whew, let's get this started:

* New entity class implementations: ElderGuardian, Husk, Stray, WitherSkeleton, ZombieVillager, AbstractHorse
* New item ids registered
* Various new API implementations from Spigot
* Tweaked storage system because of new entity id format
* New enchantment values
* Broke something with spawn eggs, will investigate

aramperes committed:
Potion color tag

aramperes committed:
Implement new entities, storage.

mastercoms commented:
There are a few dependency updates that we need to do for Glowstone 1.11.
In the pom:
`jline` to  `2.12.1`
`maven-jar-plugin` to `3.0.2`
(and maybe look into updating `maven-compiler-plugin` to `3.6.0`)
In the [Library Manager](https://github.com/GlowstoneMC/Glowstone/blob/master/src/main/java/net/glowstone/util/LibraryManager.java):
`mysql-connector-java` to `5.1.39`
`sqlite-jdbc` to `3.15.1`
Remove `commons-io`

aramperes committed:
Demo summoning spell casting

aramperes committed:
Update dependencies and libraries.

mastercoms commented on pom.xml:
```diff
@@ -77,7 +77,7 @@
             <groupId>jline</groupId>
             <artifactId>jline</artifactId>
-            <version>2.11</version>
+            <version>2.12.1</version>
```
I should mention that 2.12 had a bug with console input, might require a rewrite of the console :)

mastercoms replied:
Just kidding, that was 2.13

aramperes committed:
Entity ambient sounds, new Horse classes implementations

mastercoms commented:
A few general things that you can try to do:

* Stop using deprecated methods where it is easy to do so
* Don't use null for ItemStacks any more

aramperes commented:
@mastercoms 
I don't really know where to find such cases, in all honesty :P There are so many, and idk how to find that in IntelliJ

mastercoms commented:
Deprecated usages are listed when compiling Glowstone.

I'm not sure what to do for ItemStack, but I would look in some inventory code to start.

aramperes committed:
Remove SkeletonStore, resolve some deprecation warnings

mastercoms commented on src/main/java/net/glowstone/block/blocktype/BlockFlowerPot.java:
```diff
@@ -43,12 +42,6 @@ public TileEntity createTileEntity(GlowChunk chunk, int cx, int cy, int cz) {
         GlowBlockState state = block.getState();
-        MaterialData data = state.getData();
-
-        if (!(data instanceof FlowerPot)) {
```
That's there for old worlds, I think we should hang on to this one.

aramperes committed:
Revert FlowerPot change, first attempt at replacing null ItemStack to Material.AIR * 0 (... and broke crafting system)

aramperes committed:
Lower exhaustion decrease values

mastercoms committed:
Updated commented dependencies

mastercoms committed:
Inventory and dependency fixes

Some work on using air instead of null for empty items
Improvements to dependency list
Start work on LibraryManager checksums

mastercoms committed:
Some more work on making click handler respect empty slots

mastercoms committed:
Fix crafting to respect empty items

mastercoms committed:
Some work on making cursor and drops respect empty slots

mastercoms committed:
DROP_* drops from slot, not cursor

mastercoms committed:
Finish up click handler fixes

mastercoms committed:
Fix assert message in drop

mastercoms commented:
I think we've made all the changes that are _required_ to add 1.11 support. Of course, we still have much to do, but I think that can be done in subsequent PRs or commits. @momothereal please correct me if I'm wrong.

aramperes commented:
There are probably more tests to be done for the changes to entity storage and legacy compatibility, as well as the changes for AIRx0, but I think this is ok regarding features.

mastercoms committed:
Make armor stand respect empty items

mastercoms committed:
Improvement to click handler logic in ternary

mastercoms committed:
Merge branch 'master' into 1.11

mastercoms commented:
Considering everything looks good and changes to be made have settled a bit, I can say that the 1.11 branch is ready to be merged into master.

mastercoms committed:
Glowkit 1.11 is in master

mastercoms committed:
Merge branch '1.11' of github.com:GlowstoneMC/Glowstone into 1.11
